src/Uno.Core/Collections/ImmutableList.cs

```diff
@@ -24,33 +26,203 @@ namespace Uno.Collections
+		{
+			var itemsToAdd = items.ToArray();
+			var newData = new T[_data.Length + itemsToAdd.Length];
+			Array.Copy(_data, newData, _data.Length);
+			Array.Copy(itemsToAdd, 0, newData, _data.Length, itemsToAdd.Length);
+			return new ImmutableList<T>(newData);
+		}
+
+		public IImmutableList<T> Insert(int index, T element)
+		{
+			var newData = new T[_data.Length + 1];
+			newData[index] = element; // will throw out-of-range exception if index is not valid
+
+			if (index > 0)
+			{
+				Array.Copy(_data, newData, index);
+			}
+			if (index + 1 < _data.Length)
+			{
+				Array.Copy(_data, index, newData, index + 1, _data.Length - index);
+			}
+			return new ImmutableList<T>(newData);
+		}
+
+		public IImmutableList<T> InsertRange(int index, IEnumerable<T> items)
+		{
+			var insertedItems = items.ToArray();
+			if (insertedItems.Length == 0)
+			{
+				return this; // nothing to insert
+			}
+
+			var newData = new T[_data.Length + insertedItems.Length];
+
+			if (index > 0)
+			{
+				Array.Copy(_data, newData, index);
+			}
+
+			Array.Copy(insertedItems, 0, newData, index, insertedItems.Length);
+
+			if (index + 1 < _data.Length)
+			{
+				Array.Copy(_data, index, newData, index + insertedItems.Length, _data.Length - index);
+			}
+			return new ImmutableList<T>(newData);
+		}
+
+		public IImmutableList<T> Remove(T value, IEqualityComparer<T> equalityComparer)
+		{
+			var comparer = equalityComparer ?? EqualityComparer<T>.Default;
+			return RemoveAll(x => comparer.Equals(x, value));
+
+		}
+
+		public IImmutableList<T> RemoveAll(Predicate<T> match)
+		{
+			var newData = _data.Where(x => !match(x)).ToArray();
+			return new ImmutableList<T>(newData);
+		}
+
+		public IImmutableList<T> RemoveRange(IEnumerable<T> items, IEqualityComparer<T> equalityComparer)
+		{
+			var itemsToRemove = items.ToArray();
+			if (itemsToRemove.Length == 0)
+			{
+				return this; // nothing to remove
+			}
+			var comparer = equalityComparer ?? EqualityComparer<T>.Default;
+			return RemoveAll(x => itemsToRemove.Contains(x, comparer));
+		}
+
+		public IImmutableList<T> RemoveRange(int index, int count)
+		{
+			if (index < 0 || index >= _data.Length)
+			{
+				throw new ArgumentOutOfRangeException(nameof(index));
+			}
+			if (count < 0 || count + index > _data.Length)
+			{
+				throw new ArgumentOutOfRangeException(nameof(count));
+			}
+			if (count == 0)
+			{
+				return this;
+			}
+			if (count == _data.Length)
+			{
+				return Empty;
+			}
+
+			var newData = new T[_data.Length - count];
+			if (index > 0)
+			{
+				Array.Copy(_data, 0, newData, 0, index);
+			}
+
+			if (index + count < _data.Length)
+			{
+				Array.Copy(_data, 0, newData, 0, index);
+			}
+
+
+
+			return new ImmutableList<T>(newData);
+		}
+
+		IImmutableList<T> IImmutableList<T>.RemoveAt(int index)
+		{
+			return RemoveAt(index);
+		}
+
+		public IImmutableList<T> SetItem(int index, T value)
+		{
+			if (Equals(_data[index], value))
```

Comment:
I think here you should use ReferenceEquals

Reply:
No we need to invoke equality members there.